mesonbuild/interpreter/dependencyfallbacks.py

```diff
@@ -340,6 +340,23 @@ def lookup(self, kwargs: TYPE_nkwargs, force_fallback: bool = False) -> Dependen
                         self._subproject_impl(subp_name, varname)
                     break
 
+                # Check if we could have a matching Rust crate imported from
+                # Cargo.lock, with expected api version. We know there is no
+                # wrap-file defined to match this crate.
+                if name.endswith('-rs'):
+                    s = name.split('-')
+                    if (len(s) < 3):
+                        continue
+                    rs_ending = s.pop()
+                    version = s.pop()
+                    crate = '-'.join(s)
+                    cargo_wraps = self.wrap_resolver.get_cargo_wraps(crate)
+                    if not cargo_wraps:
+                        continue
+                    raise DependencyException(f'Rust dependency "{name}" not found, ' +
```

Comment:
DependencyException means this would be fatal, regardless of the value of required...

Reply:
Indeed. If we don't find it at this step, there is no chance we find it later, and the error message is thus confusing "Dependency X not found (tried pkgconfig and cmake)".
So I think it's better to make it an hard error instead of a simple warning.

Reply:
Someone could make this dep not required, but if it's present in the Cargo.lock file, what would be the use case?
I can check the required flag, if it makes more sense.

Reply:
The usual reason for making a dependency not required is to make it a build option so users can decide which components of the project they would like to build.

Reply:
Ok. Will check the required flag :+1:

Reply:
I would probably think to put this below in:
```python
            elif required and (dep or i == last):
                # This was the last candidate or the dependency has been cached
                # as not-found, or cached dependency version does not match,
                # otherwise func() would have returned None instead.
                raise DependencyException(f'Dependency {self._display_name!r} is required but not found.')
```

as a special-casing of the message you seek to improve. Save cargo_wraps where you currently calculate them, but emit the error message in the existing error condition.

Reply:
I would probably think to put this below in:
```python
            elif required and (dep or i == last):
                # This was the last candidate or the dependency has been cached
                # as not-found, or cached dependency version does not match,
                # otherwise func() would have returned None instead.
                raise DependencyException(f'Dependency {self._display_name!r} is required but not found.')
```

as a special-casing of the message you seek to improve. Save cargo_wraps where you currently calculate them, but emit the error message in the existing error condition.

Reply:
Sorry, but I'm not getting what this changes concretely.
This PR just tries to add an error message when a user specifiies a Rust dep, and did a mistake in version number used. We specifically want a different error message compared to the normal one.

Reply:
It should mean that if you have a C dependency with the same name it gets used as @tristan957 was worried about.

Instead of raising an error during candidate collection if "*-rs" isn't a valid crate dependency we raise an error after all dependency lookups (including both pkg-config and wraps) fail.

Reply:
(I currently have a fever. I eyeballed the code and my analysis *feels* right, but I cannot discount the possibility that what I'm saying is a fever dream in the most literal sense.)

Reply:
The problem is that cargo wrap detection is specifically done in a different place than pkg-config/cmake path.

Initially, I tried to implement something more unified, but I realized that cargo integration was modeled quite differently. So I'm not sure how to make this better.